packages/editor/src/components/table-of-contents/panel.js

```diff
@@ -5,13 +5,22 @@ import { Fragment } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 import { withSelect } from '@wordpress/data';
 
+import { delay } from 'lodash';
+
 /**
  * Internal dependencies
  */
 import WordCount from '../word-count';
 import DocumentOutline from '../document-outline';
 
-function TableOfContentsPanel( { headingCount, paragraphCount, numberOfBlocks } ) {
+function TableOfContentsPanel( { headingCount, paragraphCount, numberOfBlocks, onRequestClose } ) {
+	// Remove the block id hash from the url because block ids are only valid for the session.
+	const removeBlockIdHash = () => {
+		delay( () => {
+			const { pathname, search } = window.location;
+			window.history.pushState( '', document.title, pathname + search );
+		}, 50 );
```

Comment:
For what reason is a delay necessary? Where does the number `50` come from? What happens (should happen) if the component is unmounted between the time this is scheduled and when the callback is actually invoked?

Asynchronous behaviors in components are often a code smell, prone to unexpected consequences surrounding unanticipated component lifecycle. Can we not rely on an event to have occurred after which we can be certain to be able to remove the hash?

Reply:
the delay here is about removing the hash from the url after the browser has followed the link. In my testing this was required to successfully remove the hash, otherwise the onClick handler fires before the browser followed the link, so the hash was added after the removal attempt.

`50` is arbitrary, I can try using `defer` instead.

Reply:
changed in https://github.com/WordPress/gutenberg/pull/10815/commits/bada58fe6e0b180e36172f4800c8765bfec76766